rgw multisite secondary endpoint checks

alimaredia opened the pull request:
Add two checks in secondary.yml:

1) a check that ensures a connection to the
endpoint the realm is being fetched from.

2) a check that ensures connections from the
host rgw to all the rgws endpoints in a zone and
all the rgw endpoints for all the zonegroups
in a realm.

Closes: https://bugzilla.redhat.com/show_bug.cgi?id=1731158

Signed-off-by: Ali Maredia <amaredia@redhat.com>

cbodley commented:
conceptually this looks great :+1:

dsavineau commented on roles/ceph-rgw/tasks/multisite/secondary_connection_checks.yml:
```diff
@@ -0,0 +1,47 @@
+  - name: get zones from zonegroups list
+    set_fact:
+      zonegroups_endpoints: "{{ zonegroups_endpoints }} + {{ item.endpoints }}"
+    with_items: "{{ zonegroups }}"
```
because `zonegroups` is a dict you should use `with_dict`
This means you also need to change `item.endpoints` by `item.value.endpoints`

dsavineau commented on roles/ceph-rgw/tasks/multisite/secondary_connection_checks.yml:
```diff
@@ -0,0 +1,47 @@
+    set_fact:
+      zones: "{{ zones }} + {{ item.zones }}"
+    with_items: "{{ zonegroups }}"
+    when: item.name == "{{ rgw_zonegroup }}"
```
same here about the dict also for `item.name` it will be `item.key`

dsavineau commented on roles/ceph-rgw/tasks/multisite/secondary_connection_checks.yml:
```diff
@@ -0,0 +1,47 @@
+    set_fact:
+      zone_endpoints: "{{ zone_endpoints }} + {{ item.endpoints }}"
+    with_items: "{{ zones }}"
+    when: item.name != "{{ rgw_zone }}"
```
same about the dict

dsavineau commented on roles/ceph-rgw/tasks/multisite/secondary_connection_checks.yml:
```diff
@@ -0,0 +1,47 @@
+  - name: ensure connection to zones endpoints
+    uri:
+      url: "{{ item }}"
+    with_items: "{{ zone_endpoints }}"
```
only a minor thing but maybe we can combined both tasks

```yaml
- name: ensure connection to endpoints
  uri:
    url: "{{ item }}"
  with_items:
    - "{{ zonegroups_endpoints }}"
    - "{{ zone_endpoints }}"
```

wdyt ?

dsavineau commented:
jenkins test dev-centos-container-rgw_multisite

dsavineau commented:
jenkins test dev-centos-non_container-rgw_multisite

guits commented on roles/ceph-defaults/defaults/main.yml:
```diff
@@ -511,7 +511,7 @@ rgw_multisite_endpoint_addr: "{{ ansible_fqdn }}"
 rgw_pull_port: "{{ radosgw_civetweb_port }}"
 rgw_pull_proto: "http" # should be the same as rgw_multisite_proto for the master zone cluster
-#rgw_pullhost: localhost # rgw_pullhost only needs to be declared if there is a zone secondary. It should be the same as rgw_multisite_endpoint_addr for the master zone cluster
+#rgw_pull_host: localhost # rgw_pull_host only needs to be declared if there is a zone secondary. It should be the same as rgw_multisite_endpoint_addr for the master zone cluster
```
```suggestion
#rgw_pull_host: "{{ rgw_pullhost | default('localhost') }}" # rgw_pull_host only needs to be declared if there is a zone secondary. It should be the same as rgw_multisite_endpoint_addr for the master zone cluster
```

guits commented on roles/ceph-validate/tasks/check_rgw_multisite.yml:
```diff
@@ -55,9 +55,9 @@
   when:
     - rgw_zonesecondary
-    - rgw_pullhost is undefined
+    - rgw_pull_host is undefined
```
not sure about this condition given that you have a default value in `roles/ceph-defaults/defaults/main.yml`

alimaredia commented:
@dsavineau I changed up the commit and removed the changes to `rgw_pullhost` so all of your comments are all outdated. Here is my feedback on some of them:

All the times you wanted me to use `with_dict` instead of `with_items`, I am iterating over a list not a dict so `with_dict` doesn't apply. 

Also I'd prefer to have 2 separate checks, 1 for all the zone endpoints and 1 for all the zonegroup endpoints. This because if there is a failure the user knows that it's happening because of the connections to the zone endpoints or the zonegroup endpoints.

alimaredia commented:
@dsavineau I changed up the commit and removed the changes to `rgw_pullhost` so all of your comments are all outdated. Here is my feedback on some of them:

All the times you wanted me to use `with_dict` instead of `with_items`, I am iterating over a list not a dict so `with_dict` doesn't apply. 

Also I'd prefer to have 2 separate checks, 1 for all the zone endpoints and 1 for all the zonegroup endpoints. This because if there is a failure the user knows that it's happening because of the connections to the zone endpoints or the zonegroup endpoints.

dsavineau commented:
jenkins test centos-container-rgw_multisite

dsavineau commented:
jenkins test centos-non_container-rgw_multisite

dsavineau commented:
I need to amend the commit message because the BZ link is missing

dsavineau commented on roles/ceph-rgw/tasks/multisite/secondary_connection_checks.yml:
```diff
@@ -0,0 +1,34 @@
+
+- name: get zonegroups list from perdiod_get_json
+  set_fact:
+    zonegroups: "{{ period_get_json['period_map']['zonegroups'] }}"
```
I guess we can combine this into a single fact

dsavineau commented:
jenkins test centos-non_container-rgw_multisite

dsavineau commented:
jenkins test centos-container-rgw_multisite

dsavineau commented:
jenkins test centos-non_container-all_daemons

alimaredia commented:
@dsavineau I added these commits to https://github.com/ceph/ceph-ansible/pull/4545. Could we close this PR and move the discussion there?

dsavineau commented:
@alimaredia sure I'll close that one
